src/background.c

```diff
@@ -0,0 +1,817 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <glib/gi18n.h>
+#include <gio/gio.h>
+#include <gio/gdesktopappinfo.h>
+
+#include "background.h"
+#include "request.h"
+#include "permissions.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-utils.h"
+#include "flatpak-instance.h"
+
+/* Implementation notes:
+ *
+ * We store a YES/NO/ASK permission for "run in background".
+ *
+ * There is a portal api for apps to request this permission
+ * ahead of time. The portal also lets apps ask for being
+ * autostarted.
+ *
+ * We determine this condition by getting per-application
+ * state from the compositor, and comparing that list to
+ * the list of running flatpak instances obtained from
+ * $XDG_RUNTIME_DIR/.flatpak/. A thread is comparing
+ * this list every minute, and if it finds an app that
+ * is in the background twice, we take actions:
+ * - if the permission is NO, we kill it
+ * - if the permission is YES or ASK, we notify the user
+ *
+ * We only notify once per running instance to not be
+ * annoying.
+ *
+ * Platform-dependent parts are in the background portal
+ * backend:
+ * - Notifying the user
+ * - Getting compositor state
+ * - Enable or disable autostart
+ */
+
+#define PERMISSION_TABLE "background"
+#define PERMISSION_ID "background"
+
+typedef struct _Background Background;
+typedef struct _BackgroundClass BackgroundClass;
+
+struct _Background
+{
+  XdpBackgroundSkeleton parent_instance;
+};
+
+struct _BackgroundClass
+{
+  XdpBackgroundSkeletonClass parent_class;
+};
+
+static XdpImplAccess *access_impl;
+static XdpImplBackground *background_impl;
+static Background *background;
+
+GType background_get_type (void) G_GNUC_CONST;
+static void background_iface_init (XdpBackgroundIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (Background, background, XDP_TYPE_BACKGROUND_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_TYPE_BACKGROUND, background_iface_init));
+
+
+typedef enum { UNSET, NO, YES, ASK } Permission;
+
+static GVariant *
+get_all_permissions (void)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) out_perms = NULL;
+  g_autoptr(GVariant) out_data = NULL;
+
+  if (!xdp_impl_permission_store_call_lookup_sync (get_permission_store (),
```

Comment:
It would be nice if this used a utility that used notifications to get a constantly up-to-date GVariant for a permission query  without having to actually query it all the time. However, that can be done separately.

Reply:
agreed